### ISSUE TITLE
Out of memory when loading a corrupted vlog

### ISSUE THREAD

**Issue by peterstace** [dgraph-io/badger#154]
ALICE detected that when a corrupt (garbage append) vlog is opened, Badger crashes. The entry header has garbage in it, which can encode erroneously 'large' key/value lengths. This causes allocation of slices that don't fit in memory.

E.g.:

```
0       stdout("'start:big\n'")
1       creat("LOCK", parent=1355754, mode='0666', inode=1359282)                                                                                               2       append("LOCK", offset=0, count=6, inode=1359282)
3       creat("MANIFEST", parent=1355754, mode='0666', inode=1359283)
4       trunc("MANIFEST", initial_size=0, inode=1359283, final_size=0)
5       fsync(".", size=100, inode=1355754)
6       creat("000000.vlog", parent=1355754, mode='0666', inode=1359288)
7       fsync(".", size=120, inode=1355754)
8       append("000000.vlog", offset=0, count=1048670, inode=1359288)
9       append("000000.vlog", offset=1048670, count=1048670, inode=1359288)
10      append("000000.vlog", offset=2097340, count=1048670, inode=1359288)
11      append("000000.vlog", offset=3146010, count=1048670, inode=1359288)
12      append("000000.vlog", offset=4194680, count=1048670, inode=1359288)
13      creat("000001.vlog", parent=1355754, mode='0666', inode=1359291)
14      fsync(".", size=140, inode=1355754)
15      append("000001.vlog", offset=0, count=1048670, inode=1359291)
16      creat("000001.sst", parent=1355754, mode='0666', inode=1359294)
17      fsync(".", size=160, inode=1355754)
18      append("000001.sst", offset=0, count=470, inode=1359294)
19      file_sync_range("000001.sst", count=470, offset=0, inode=1359294)
20      append("MANIFEST", offset=0, count=8, inode=1359283)
21      fsync("MANIFEST", size=8, inode=1359283)
22      unlink("LOCK", parent=1355754, inode=1359282, size=6, hardlinks=1)
23      fsync(".", size=160, inode=1355754)
24      fsync(".", size=160, inode=1355754)
25      stdout("'stop:big\n'")
-------------------------------------
Finding vulnerabilities...
...
(Dynamic vulnerability) Atomicity: Operation 8(???)
(Dynamic vulnerability) Atomicity: Operation 9(garbage written semi-expanded (3 count splits), garbage written)
(Dynamic vulnerability) Atomicity: Operation 10(???)
(Dynamic vulnerability) Atomicity: Operation 11(garbage written)
(Dynamic vulnerability) Atomicity: Operation 12(???)
(Dynamic vulnerability) Atomicity: Operation 15(???)
...
```

**Comment by peterstace** [comment]
A possible solution would be to add a CRC after the header. Don't try to allocate for the key/value if the hash is no good. We'll still need a CRC after the key and value to validate that they're ok.

**Comment by manishrjain** [comment]
Just ensure that the value size is not more than 1G. We can easily add that check when accepting key-values and similarly when replaying.

**Comment by peterstace** [comment]
I'll put max size for keys to be 1G as well. Although I don't think we'll even need keys that big, this is really more about preventing out of memory crashes rather than restricting users.

**Comment by peterstace** [comment]
Fixed in #158
